### PR TITLE
refactor(storage): Use TableKey struct in the state store interfaces

### LINKED ISSUE #11927: Use `TableKey` struct in memtable instead of `Bytes`
This requires a bigger refactoring since the memtable key is in `Bytes` not `TableKey` type. Given that the memtable is already in storage, I will try to expose the `TableKey` struct in memtable as well in a separate PR and fix this issue then.

_Originally posted by @hzxa21 in https://github.com/risingwavelabs/risingwave/pull/11880#discussion_r1308165072_

### PR DESCRIPTION
**Pull request by hzxa21**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Fix #11927

- Use TableKey struct in the state store interfaces
- Remove unused codes (src/storage/src/hummock/store/memtable.rs, src/storage/src/write_batch.rs)
- Move `HummockStorage` into src/storage/hummock/store/hummock_storage.rs
- Move `LocalHummockStorage` into src/storage/hummock/store/local_hummock_storage.rs

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/storage/src/mem_table.rs [hunk 1]
```diff
@@ -46,15 +51,19 @@ pub enum KeyOp {
 /// `MemTable` is a buffer for modify operations without encoding
 #[derive(Clone)]
 pub struct MemTable {
-    pub(crate) buffer: BTreeMap<Bytes, KeyOp>,
+    pub(crate) buffer: BTreeMap<TableKey<Bytes>, KeyOp>,
```
Comment (wcy-fdu) [inline src/storage/src/mem_table.rs:54]: Do we need `#[derive(EstimateSize)]` for TableKey to make memory statistics more accurate?

Comment (hzxa21) [reply src/storage/src/mem_table.rs:54]: It is already done in this PR [here](https://github.com/risingwavelabs/risingwave/pull/11943/files#diff-e8546b48ec0e07c8f2c4d526da0e83dc956d44ec12929864691d9b82ed06418aR415).

**Review by wenym1 — APPROVED** [review]

File: src/storage/clippy.toml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+disallowed-methods = [
```
Comment (Li0k) [inline src/storage/clippy.toml:1]: `clippy.toml` is expected ?

**Review by Li0k — APPROVED** [review]
LGTM

File: src/storage/clippy.toml (hunk 2 above)
Comment (hzxa21) [reply src/storage/clippy.toml:1]: Yes, I need to add `ignore-interior-mutability = ["risingwave_hummock_sdk::key::TableKey"]` so I just copy rules from src/streaming/clippy.toml as well.

**Pull request merged**